### PR TITLE
include complete redirect rule feature

### PR DESCRIPTION
**Pull request by x0day**
hi, thanks for this project first, it helps me a lot.

in this PR, I complete some feature about scriptlet injection and redirect rules.

1. update adblock-rust => 0.4.3
2. add `include_redirect_urls` parameter for `FilterSet.add_filter_list` and `FilterSet.add_filters`
3. add `Engine.add_resource` method.
4. add `adblock.AddResourceError` Exception
5. add `redirect_type` for `BlockerResult`
6. add some tests for these functions.

### CODE REVIEW

**Comment by x0day** [comment]
Some implementations refer to the https://github.com/brave/adblock-rust-ffi.

File: src/lib.rs [hunk 1]
```diff
@@ -102,13 +113,30 @@ pub struct BlockerResult {
 
 impl From<RustBlockerResult> for BlockerResult {
     fn from(br: RustBlockerResult) -> Self {
+        let mut redirect: Option<String> = None;
+        let mut redirect_type: Option<String> = None;
+        if br.redirect.is_some() {
+            let resource = br.redirect.unwrap();
```
Comment (ArniDagur) [inline src/lib.rs:119]: Can you refactor this to get rid of the unwrap, and even the mutability? I'm thinking something like
```rust
let (redirect, redirect_type) = if let Some(resource) = br.redirect {
    // Code here...
} else {
    (None, None)
}
```

**Review by ArniDagur — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 2]
```diff
@@ -446,6 +497,36 @@ impl Engine {
...
+            kind: ResourceType::Mime(MimeType::from(std::borrow::Cow::from(
+                content_type.to_string(),
+            ))),
+            content: content.to_string(),
+        });
+
+        match result {
+            Ok(_) => Ok(()),
+            Err(err) => match err {
+                RustAddResourceError::InvalidBase64Content => Err(AddResourceError::new_err(
```
[the remark this answers is not in the document]
Comment (x0day) [reply src/lib.rs:520]: you are right, I also want do this as you think, but like the question above, I can't find any test case for `RustAddResourceError::InvalidUtf8Content` emmm..
so I left it with `AddResourceError`. so let me changed it.

File: tests/test_exceptions.py [hunk 3]
```diff
@@ -9,3 +10,14 @@ def test_correct_baseclasses():
...
+
+
+def test_add_resource_error():
+    filter_set = adblock.FilterSet()
+    engine = adblock.Engine(filter_set=filter_set)
+
+    with pytest.raises(adblock.AddResourceError) as exc:
+        engine.add_resource(name="aa", content_type="image/jpeg", content="111")
+
+    assert "invalid base64 content" in str(exc.value)
```
[the remark this answers is not in the document]
Comment (x0day) [reply tests/test_exceptions.py:24]: https://github.com/brave/adblock-rust/blob/2288ff9e140272269f433f1a739f128fcd74e755/src/resources/mod.rs#L151

adblock will ensure any text contents are also valid utf8 when content_type is javascript or text. I added a test for it.

**Comment by x0day** [comment]
@ArniDagur  still have something not ready to merge?

**Comment by x0day** [comment]
> Thank you for your contribution. This has been merged as https://github.com/ArniDagur/python-adblock/pull/60.

can you release a new version on pypi.org? thank you.

**Pull request closed without merging**